zerver/models.py

```diff
@@ -1423,20 +1423,43 @@ def get_realm_playgrounds(realm: Realm) -> List[RealmPlaygroundDict]:
     return playgrounds
 
 
-# The Recipient table is used to map Messages to the set of users who
-# received the message.  It is implemented as a set of triples (id,
-# type_id, type). We have 3 types of recipients: Huddles (for group
-# private messages), UserProfiles (for 1:1 private messages), and
-# Streams. The recipient table maps a globally unique recipient id
-# (used by the Message table) to the type-specific unique id (the
-# stream id, user_profile id, or huddle id).
 class Recipient(models.Model):
+    """A Recipient object represents an audience that can potentially
+    receive messages in Zulip.
+
+    This table essentially functions as a generic foreign key that
+    allows Message.recipient_id to be a simple ForeignKey representing
+    the audience for a message, while supporting the different types
+    of audiences Zulip supports for a message.
+
+    Recipient has just two attributes: The enum type, and a type_id,
+    which is the ID of the UserProfile/Stream/Huddle object containing
+    all the metadata for the audience. There are 3 recipient types:
+
+    1. 1:1 private message: The type_id is the ID of the UserProfile
+       who will receive any message to this Recipient. The sender is
+       of such a message is represented separately.
+    2. Stream message: The type_id is the ID of the associated Stream.
+    3. Group private message: In Zulip, group private messages are
+       represented by Huddle objects, which encode the set of users
+       in the conversation. The type_id is the ID of the associated Huddle
+       object; the set of users is usually retrieved via the Subscription
+       table. See the Huddle class for details.
+
+    The actual audience is encoded in the Subscription table.
```

Comment:
This sentence feels confusing to me. Was it supposed to be part of (3) or just separate, talking in general? By audience, I presume we mean something like "the set of users meant to receive the message", right? Though that sentence wouldn't apply to 1:1 PMs  I guess, because that audience is specified by type_id already, without needing to think about Subscription.

Reply:
Even for 1:1PMs, there's a Subscription table row that records which user is subscribed to that Recipient object; but I guess we don't even use that in the message send code path.

Reply:
Replaced this with a cleaner sentence.